library/model/library.py

```diff
@@ -0,0 +1,154 @@
+            return 5
+        elif isinstance(self.book, ElectronicBook):
+            return 2
+        elif isinstance(self.book, AudioBook):
+            return 2
+        else:
+            raise AttributeError("No such book type...")
+
+class BorrowedLibraryBook:
+    borrowed_by: User
+    listing: LibraryBookListing
+    due_date: datetime
+    current_fee: float
+    was_returned: bool
+
+    def __init__(self, borrowed_by: User, listing: LibraryBookListing, due_date: datetime, current_fee: float):
+        self.borrowed_by = borrowed_by
+        self.listing = listing
+        self.due_date = due_date
+        self.current_fee = current_fee
+        self.was_returned = False
+
+class LibraryAccount:
+    user: User
+    borrowed_books: list[BorrowedLibraryBook]
+    pending_payment: list[BorrowedLibraryBook]
+    read_books: list[Book]
+    reading_credits: int
+
+    def __init__(self, user: User):
+        self.user = user
+        self.borrow_books = []
+        self.pending_payment = []
+        self.read_books = []
+
+class LibraryAccountInvoice:
+    user: User
+    borrows: list[BorrowedLibraryBook]
+    fee: float
+
+    def __init__(self, user: User, borrows: list[BorrowedLibraryBook], fee: float):
+        self.user = user
+        self.borrows = borrows
+        self.fee = fee
+
+class Library:
+    accounts: dict[str, LibraryAccount]
+    books: dict[str, LibraryBookListing]
+
+    def __init__(self):
+        self.books = {}
+
+    def create_listing(self, book: Book):
+        if book.isbn not in self.books:
+            self.books[book.isbn] = LibraryBookListing(book)
+        else:
+            self.books[book.isbn].existing_items += 1
+
+    def renew_rental(self, book: BorrowedLibraryBook):
+        book.due_date += timedelta(days=7)
+        book.current_fee += book.listing.get_weekly_fee()
+        return self
+
+    def return_book(self, borrowed: BorrowedLibraryBook):
+        if borrowed.was_returned:
+            raise ValueError("book was already returned")
+        if borrowed.listing.book_type == "Paper":
+            self.borrowed_items -= 1
+        account = self.accounts[borrowed.borrowed_by.email]
+        account.read_books.append(borrowed.listing.book)
+        account.pending_payment.append(borrowed)
+        account.borrowed_books.remove(borrowed)
+        borrowed.was_returned = True
+
+    def borrow_book(self, listing: LibraryBookListing, user: User):
+        if not user.email in self.accounts:
+            raise ValueError("Unknown user")
+        if not listing.can_be_borrowed():
+            raise ValueError("Book cannot be borrowed")
+        if listing.book_type == "Paper":
+            listing.borrowed_items += 1
+        due_date = datetime.now() + timedelta(days=7)
+        current_fee = self.get_weekly_fee()
+        return BorrowedLibraryBook(user, listing, due_date, current_fee)
+
+    @staticmethod
+    def compute_reading_credits(borrows: list[BorrowedLibraryBook]) -> int:
```

Comment:
a static method seems a bit out of place here; not sure if this belongs to the library class